packages/block-library/src/post-author/edit.js

```diff
@@ -1,25 +1,263 @@
+/**
+ * External dependencies
+ */
+import classnames from 'classnames';
 /**
  * WordPress dependencies
  */
-import { useEntityProp, useEntityId } from '@wordpress/core-data';
-import { useSelect } from '@wordpress/data';
-import { sprintf, __ } from '@wordpress/i18n';
-
-function PostAuthorDisplay() {
-	const [ authorId ] = useEntityProp( 'postType', 'post', 'author' );
-	const author = useSelect(
-		( select ) =>
-			select( 'core' ).getEntityRecord( 'root', 'user', authorId ),
-		[ authorId ]
+import apiFetch from '@wordpress/api-fetch';
+import { useRef } from '@wordpress/element';
+import { useEntityProp } from '@wordpress/core-data';
+import {
+	AlignmentToolbar,
+	BlockControls,
+	FontSizePicker,
+	InspectorControls,
+	RichText,
+	__experimentalUseColors,
+	withFontSizes,
+} from '@wordpress/block-editor';
+import { PanelBody, SelectControl, ToggleControl } from '@wordpress/components';
+import { useSelect, useDispatch } from '@wordpress/data';
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import BlockColorsStyleSelector from './block-colors-selector';
+
+const DEFAULT_AVATAR_SIZE = 24;
+
+function PostAuthorDisplay( { props, author, authors } ) {
+	const ref = useRef();
+
+	const { editPost } = useDispatch( 'core/editor' );
+
+	const { fontSize, setFontSize } = props;
+	const {
+		TextColor,
+		BackgroundColor,
+		InspectorControlsColorPanel,
+		ColorPanel,
+	} = __experimentalUseColors(
+		[
+			{ name: 'textColor', property: 'color' },
+			{ name: 'backgroundColor', className: 'background-color' },
+		],
+		{
+			contrastCheckers: [
+				{
+					backgroundColor: true,
+					textColor: true,
+					fontSize: fontSize.size,
+				},
+			],
+			colorDetector: { targetRef: ref },
+			colorPanelProps: {
+				initialOpen: true,
+			},
+		},
+		[ fontSize.size ]
+	);
+
+	const { align, id, showAvatar, showBio, byline } = props.attributes;
+
+	const avatarSizes = [
+		{ value: 24, label: __( 'Small' ) },
+		{ value: 48, label: __( 'Medium' ) },
+		{ value: 96, label: __( 'Large' ) },
+	];
+
+	const changeAuthor = ( authorId ) => {
+		apiFetch( { path: '/wp/v2/users/' + authorId + '?context=edit' } ).then(
```

Comment:
I think there's a race condition here, let's imagine:

* I change the author and this request is slow and just hangs
* I change the author again and request succeeds immediately, `setAttributes()` is called for that author
* Now the first request finally succeeded, `setAttributes()` is called for that author which is not what we wanted

Should this be wrapped in an effect that cancels any existing request when authorId changes?

Reply:
yes we should and I will :)

Reply:
It now uses an effect to fetch the data and I also return a cancel function that aborts all calls.

Reply:
This is because once the fetch callback is executed, `<PostAuthorDisplay/>` is unmounted and then mounted again which causes the effect to run twice. The following check from `<PostAuthorEdit>` is responsible:

```js
if ( ! author ) {
 return 'Post Author Placeholder';
}
```

Reply:
By "this" you mean that we have two requests for one author change :)